docs/design/proposals/creds-from-kms.md

```diff
@@ -0,0 +1,130 @@
+# Fetching Ceph user credentials from KMS
+
+## Proposal
+
+Ceph-CSI supports several KMS implementations for storing and
+fetching DEKs (Data Encryption Keys) used for RBD volume encryption.
+The focus of this proposal is to leverage existing KMS implementation
+to support obtaining Ceph user keyring used while
+creating, deleting, mapping and resizing RBD volumes.
+
+### Benefits
+
+- For scenarios where Ceph is deployed externally, either as
+  a standalone or through Rook, one needs to obtain the Ceph keyring(s)
+  and manually create the K8s secrets needed by Ceph-CSI. With this feature,
+  sensitive keyring(s) can be stored securely in an
+  external secret management system such as Hashicorp Vault,
+  AWS Secret Manager, etc., and have Ceph-CSI pull these
+  keys directly from them.
+- More secure compared to K8s secrets,
+  which are relatively easy to access and
+  store sensitive secrets as unencrypted based64 encoded text.
+
+### Drawbacks
+
+- Adds additional overhead as on each CSI RPC call
+  a request is made to KMS for fetching credentials.
+- Possible risk of hitting the KMS rate limit
+
+## Extending existing implementation
+
+- KMS ID is  provided as part of the provisioner,
+  node-stage, and controller-expand secrets, along with Ceph `userID`, as shown below.
+  Having this information as part of *StorageClass* parameters,
+  like using the `encryptionKMSID` key for volume encryption,
+  wouldn't work as these parameters are not passed
+  to all CSI RPCs, e.g. `DeleteVolume`.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: csi-rbd-provisioner
+stringData:
+  userID: <ceph-user-id>
+  kmsID: <kms-id>
+ ```
+
+- A new ConfigMap with default name `ceph-csi-creds-kms-config` similar to
+  [ceph-csi-encryption-kms-config](https://github.com/ceph/ceph-csi/blob/devel/examples/kms/vault/kms-config.yaml)
+  will be added.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+data:
+    config.json: |-
+        {<kms-id>: <config>}
+metadata:
+    name: ceph-csi-creds-kms-config
+```
+
+- For KMS that rely on tenants' namespace to obtain required
```

Comment:
this is not something specific to this feature right?

Reply:
Sorry, I don't follow the question completely. 

The need to add "tenantNamespace" in config is specific to this feature, as we don't have the tenant namespace available in all RPC calls. RBD volume encryption uses the tenants' namespace for all its operations.

Reply:
add the new field to the existing struct and reuse it, lets avoid too many configurations

Reply:
Makes sense. I updated the section.